km/km_filesys.c

```diff
@@ -1864,32 +1872,71 @@ size_t km_fs_core_notes_length()
             }
          } else if (file->sockinfo == NULL) {
+            queuedbytes = 0;
+            if (file->how == KM_FILE_HOW_PIPE_1) {
+               // We are looking at the write end of a pipe, find out how much data is queued
+               queuedbytes = ioctlfionread(file->ofd);
+            }
             ret += km_note_header_size(KM_NT_NAME) + sizeof(km_nt_file_t) +
-                   km_nt_file_padded_size(file->name);
+                   km_nt_file_padded_size(file->name) + km_nt_chunk_roundup(queuedbytes);
          } else {
+            queuedbytes = 0;
+            if (file->how == KM_FILE_HOW_SOCKETPAIR0 || file->how == KM_FILE_HOW_SOCKETPAIR1) {
+               queuedbytes = ioctlfionread(file->ofd);
+            }
             ret += km_note_header_size(KM_NT_NAME) + sizeof(km_nt_socket_t) +
-                   km_nt_file_padded_size(file->name);
+                   km_nt_file_padded_size(file->name) + km_nt_chunk_roundup(queuedbytes);
          }
       }
    }
    return ret;
 }
 
+static inline size_t
+fs_core_save_pipe_contents(char* buf, size_t length, km_file_t* file, int writefd, size_t queuedbytes)
+{
+   km_assert(queuedbytes <= length);
+
+   // Read the bytes queued in the pipe
+   ssize_t bytesread = read(file->ofd, buf, queuedbytes);
+   if (bytesread < 0) {
+      km_err(1, "read pipe fd %d, %ld bytes failed", file->ofd, queuedbytes);
+   }
+   if (bytesread != queuedbytes) {
+      km_errx(1, "fd %d: expected to read %ld bytes, got %ld bytes", file->ofd, queuedbytes, bytesread);
+   }
+   // now write the bytes back into the pipe
+   ssize_t byteswritten = write(writefd, buf, bytesread);
+   if (byteswritten < 0) {
+      km_err(1, "write pipe fd %d, %ld bytes failed", writefd, bytesread);
+   }
+   if (byteswritten < bytesread) {
+      km_errx(1, "fd %d, expected to write %ld bytes, wrote %ld bytes", writefd, bytesread, byteswritten);
+   }
+   return byteswritten;
+}
+
 static inline size_t fs_core_write_nonsocket(char* buf, size_t length, km_file_t* file, int fd)
 {
    struct stat st = {};
    if (fstat(fd, &st) < 0) {
       km_warn("fstat failed fd=%d errno=%d - ignore", fd, errno);
    }
 
-   km_infox(KM_TRACE_SNAPSHOT, "fd=%d %s", fd, file->name);
+   size_t queuedbytes = 0;
+   if (file->how == KM_FILE_HOW_PIPE_1) {
+      // This is the write side of the pipe, get the number of bytes waiting on the read end.
+      queuedbytes = ioctlfionread(file->ofd);
+   }
+   km_infox(KM_TRACE_SNAPSHOT, "fd=%d %s, queuedbytes %ld", fd, file->name, queuedbytes);
    char* cur = buf;
    size_t remain = length;
    cur += km_add_note_header(cur,
                              remain,
                              KM_NT_NAME,
                              NT_KM_FILE,
-                             sizeof(km_nt_file_t) + km_nt_file_padded_size(file->name));
+                             sizeof(km_nt_file_t) + km_nt_file_padded_size(file->name) +
+                                 roundup(queuedbytes, 4));
```

Comment:
`km_nt_chunk_roundup`? Also what about the `= km_nt_chunk_roundup(fnote->datalength);` down below on line 1957?

Reply:
Will fix.